### ISSUE TITLE
Possibly incorrect implementation of the MOD opcode

### ISSUE THREAD

**Issue by joanbm** [tykel/mash16#7]
I was upgrading my own emulator to add support for the Chip16 v1.3 instructions, and the new division instructions seemed ambiguous to me, specially because behavior around negative operands tends to be problematic, and because there were both MOD and REM instructions without a clear difference between them. Not even the sign of the division operation is not even clearly defined for C89 and C++03, unless you use the div() function (See https://en.wikipedia.org/wiki/Modulo_operation), and information about what the differences between MOD and REM should be online is pretty sparse.

So I wrote a [Test ROM](https://bitbucket.org/joanbrugueram/chip16mul/raw/2ded0b85a67dafbb1cbf7b051f6f5e3c2af06feb/TestROMs/Opsv13.asm) and tested it in the two most famous emulators, mash16 and RefChip16. I found that the DIV and REM operations were implemented consistently between the two emulators (they were just the regular results of the `/` and `%` operator you get with GCC). However, I noticed that there was a different behavior on those two emulators on the MOD operation when the sign of the divisor was negative (the tests labeled `test43` and `test44` in the test ROM).

There is a pretty detailed write-up on the Test ROM header on the exact behavior implemented by both emulators. While the Chip16 reference is vague about how the instructions should behave, in my opinion I think that mash16's behavior is much harder to justify, both because it's different than other implementations of the modulo such as Haskell's one which are designed around certain useful mathematical properties, and also because behavior like -2 MOD -4 = -6 seems pretty unintuitive.

If you want to play around with it, the [Test ROM is also in .c16 format here](https://bitbucket.org/joanbrugueram/chip16mul/raw/2ded0b85a67dafbb1cbf7b051f6f5e3c2af06feb/TestROMs/Opsv13.c16). It should display a green screen if the MOD behavior is the same as RefChip16. Currently it displays a red screen with Mash16.

Would be happy to see your thoughts about this!

**Comment by tykel** [comment]
Hello Joan, I would not be surprised if this were the case. The Chip16 spec was not sufficiently detailed in quite a few aspects.

I will take a closer look at the behavior in this emulator, thanks for bringing it to my attention.

**Comment by joanbm** [comment]
FWIW I fixed the links to the test ROM that broke for some reason.

**Comment by tykel** [comment]
I just tried your test ROM and I got a green pass screen with both interpreter and JIT mode (--cpu-rec).
So I guess enough has changed since you filed this issue that I fixed it in the meantime! :)

**Comment by tykel** [comment]
Some post-mortem as this intrigued me.
I looked a bit closer and back in December 2013 (commit 7ee4511685789906ca9f891fb04373e17bf95119) I changed the definition of `mod` for ops A3, A4, A5 to be:

```
#define mod(x, y) ((((x)%(y))+(y))%(y))
```

which does satisfy `2 % -4 == -2` and `-2 % -4 == -2` as per test43 and test44 of your ROM.

Given C and C++ do truncated division per the Modulo Wikipedia page; we know that, for a dividend `x` and divider `y`:

- For x > 0, y < 0, `z = x % y` > 0, with `z in ]y .. 0]`.
Then, `(z + y) in ]0 .. y]` [1].
And `((z + y) % y) > 0` per truncated division as per [1].
- Similarly for x < 0, y < 0.

I will be putting out a new revision of the Chip16 spec in the near future, and will clarify this.

**Comment by joanbm** [comment]
Yeah, I actually did a quick bisect today and noticed that the bug I was reporting was in the last released version (i.e. v0.5.10) but not in what was most likely the latest commit at the time I created the issue, so maybe it was bogus all along.

But at least, it's good that the behaviour of MOD/REM is getting a more clear cut definition since there's certainly a lot of room for ambiguity / divergent implementations.
